README
This is the README file for stmflasher. Installation
instructions may be found in the INSTALL file.

stmflasher is Copyright (C) 2010-2013 by:

Geoffrey McRae <geoff@spacevs.com>,
Steve Markgraf <steve@steve-m.de>,
Tormod Volden,
Alatar <alatar_@list.ru>,
and others.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version. See the file COPYING for more details.

About this software
-------------------

STM Flasher is a simple command-line tool for download/upload firmware to/from
STM32 MCU and manipulate MCU flash protection with factory preprogrammed serial
bootloader. This project is a fork of stm32flash by Geoffrey McRae, hosted on
google code - http://code.google.com/p/stm32flash/.

Main features:
* device type identification
* write to flash/ram
* read from flash/ram
* auto-detect Intel HEX or raw binary input format with option to force binary
* save flash/ram block to binary file
* verify & retry up to N times on failed writes
* enable/disable flash read protection
* disable flash write protection
* start execution at specified address (-g)
* software reset the device when finished if -g not specified
* automatic resume already initialized connection (for when reset fails)
* automatic retry to send INIT cmd, if no answer from bootloader
* verbose and silent modes
* work on POSIX systems (Linux, FreeBSD, MacOS X, etc) and Windows.

Supported chips:
* 0x412 - STM32 Low-density
* 0x410 - STM32 Medium-density
* 0x414 - STM32 High-density
* 0x418 - STM32 Connectivity line
* 0x420 - STM32 Low/Medium-density VL
* 0x428 - STM32 High-density VL
* 0x430 - STM32 XL-density
* 0x416 - STM32L Medium-density
* 0x427 - STM32L Medium-density PL (?)
* 0x436 - STM32L High-density
* 0x411 - STM32F2 (Limited support)
* 0x413 - STM32F4 (Limited support)
* 0x440 - STM32F051
* 0x444 - STM32F050 (Low-density) (?)
* 0x422 - STM32F30 & F31 (?)
* 0x432 - STM32F37 & F38 (?)

Note that on STM32F2/F4 only first 3 pages (48k of flash) can be used currectly.
Use FullErase flag (-E) if you need to write to highest pages.
(?) - These are not (yet) in AN2606, so reserved by bootloader memory not known.

Usage
-----

stmflasher -p ser_port [-b rate] [-EvMKfc] [-S address[:length]] [-s start_page[:n_pages]]
        [-n count] [-r|w filename] [-ujkeiR] [-g address] [-V level] [-h]

        -p ser_port     Serial port name
        -b ser_port     Serial port baud rate (default 57600)

        -r filename     Read flash to file (stdout if "-")
        -w filename     Write flash from file (stdin if "-")
        -u              Disable the flash write-protection
        -j              Enable the flash read-protection
        -k              Disable the flash read-protection
        -e              Erase only
        -g address      Start execution at specified address (0 = flash start)
        -i              Print information about target device and serial mode
        -R              Reset controller (default for read/write/erase/etc)

        -E              Full erase
        -v              Verify writes
        -n count        Retry failed writes up to count times (default 10)
        -S address[:length]     Specify start address and optionally length for
                                read/write/erase operations
        -s start_page[:n_pages] Specify start address at page <start_page> (0 = flash start)
                                and  optionally number of pages to erase
        -M              Work with RAM, not Flash (read/write/erase operation)
        -K              Don`t Reset controller after operation (keep in bootloader)
        -f              Force binary parser
        -c              Resume the connection (don't send initial INIT)
                        *Baud rate must be kept the same as the first init*
                        This is useful with -K or if the reset fails
        -V level        Verbose output level (0 - silent, 1 - default, 2 - debug)

        -h              Show this help

Examples:
        Get device information:
                ./stmflasher -p /dev/ttyS0 -i
        Write with verify and then start execution:
                ./stmflasher -p /dev/ttyS0 -w filename -v -g 0x0
        Show information and read flash to file:
                ./stmflasher -p /dev/ttyS0 -i -r filename
        Read 100 bytes of RAM with offset 0x1000 to stdout:
                ./stmflasher -p /dev/ttyS0 -r - -M -S 0x1000:100
        Read first page of flash to file in verbose mode:
                ./stmflasher -p /dev/ttyS0 -r readed.bin -S :1 -V
        Start execution:
                ./stmflasher -p /dev/ttyS0 -g 0x0


NEWS
stmflasher v0.6.2          07.03.2013

 * Fixed validating of execution address in case when it is not
   specified (tanks to Tim Yamin)

stmflasher v0.6.1          27.02.2013

 + Allow to specify relation execution address

stmflasher v0.6.0          26.02.2013

 + Allow to work with EEPROM on STM32L
 + Chang way to use relative addressation (use '+' sign)
 * Re-factoring of calculation working space
 * Fixed soft reset, broken in 0.5.1

stmflasher v0.5.1          24.02.2013

 + Added STM32L Medium-density Plus (id 0x427) and STM32F050x (0x444)
 + Added info about EEPROM for STM32L

stmflasher v0.5            21.02.2013

 First release since fork. Changes from stm32flash SVN r61.
 * applied patches ("merging branch" on gitorious, thanks to Tormod Volden)
 ** issue10 - fixed "More than two bytes sent"
 ** issue16 - small fixes for win32
 ** issue17 - additional serial speeds
 ** issue22 - support for reading from stdin/writing to stdout
 ** issue29 - support STM32F0 devices
 ** issue31 - support for Readout Unprotect command
 ** issue32 - corrected help information
 ** issue34 - new Reset code, STM32F3 & STM32F0 support, small improvements
 * fixes for merging branch
 ** fixed memory write procedure when we write file to default location
 ** fixed memory erase proc for chips with Extended Erase (with more than 255
    flash pages)
 * major redesign of command line interface
 * added separate flag to work with RAM
 * allow to plug/reset device after flasher is started (5 INIT retries)
 * fixed "Failed to read byte: read_byte: Success"
 * set the same timeout on Win32 and Posix (3 sec)
 * fixed parity check on Posix
 * many other internal fixes


stm32.c
/*
  stmflasher - Open Source ST MCU flash program for *nix
  Copyright (C) 2010 Geoffrey McRae <geoff@spacevs.com>
  Copyright (C) 2011 Steve Markgraf <steve@steve-m.de>
  Copyright (C) 2012 Tormod Volden
  Copyright (C) 2012-2013 Alatar <alatar_@list.ru>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <stdlib.h>
#include <assert.h>
#include <stdio.h>
#include <string.h>

#include "stm32.h"
#include "utils.h"

#define STM32_ACK	0x79
#define STM32_NACK	0x1F
#define STM32_CMD_INIT	0x7F
#define STM32_CMD_GET	0x00	/* get the version and command supported */
#define STM32_CMD_EE	0x44	/* extended erase */

struct stm32_cmd {
	uint8_t get;
	uint8_t gvr;
	uint8_t gid;
	uint8_t rm;
	uint8_t go;
	uint8_t wm;
	uint8_t er; /* this may be extended erase */
	uint8_t wp;
	uint8_t uw;
	uint8_t rp;
	uint8_t ur;
};

/* Reset code for ARMv7-M (Cortex-M3) and ARMv6-M (Cortex-M0)
 * see ARMv7-M or ARMv6-M Architecture Reference Manual (table B3-8)
 * or "The definitive guide to the ARM Cortex-M3", section 14.4.
 */
static const uint8_t stm_reset_code[] = {
	0x01, 0x49,		// ldr     r1, [pc, #4] ; (<AIRCR_OFFSET>)
	0x02, 0x4A,		// ldr     r2, [pc, #8] ; (<AIRCR_RESET_VALUE>)
	0x0A, 0x60,		// str     r2, [r1, #0]
	0xfe, 0xe7,		// endless: b endless
	0x0c, 0xed, 0x00, 0xe0,	// .word 0xe000ed0c <AIRCR_OFFSET> = NVIC AIRCR register address
	0x04, 0x00, 0xfa, 0x05	// .word 0x05fa0004 <AIRCR_RESET_VALUE> = VECTKEY | SYSRESETREQ
};

static const uint32_t stm_reset_code_length = sizeof(stm_reset_code);

/* Device table, corresponds to the "Bootloader device-dependant parameters"
 * table in ST document AN2606.
 * Note that the option bytes upper range is inclusive!
 */
const stm32_dev_t devices[] = {
//	{ PID ,         NAME                   , RAM start , RAM bl res, RAM end   ,FLASH start, FLASH end ,pps, psize, Mem start , Mem end   , Opt start ,  Opt end  ,EEPROM start,EEPROM end},
	{0x412, "STM32F Low-density"           , 0x20000000, 0x20000200, 0x20002800, 0x08000000, 0x08008000,  4, 1024 , 0x1FFFF000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x410, "STM32F Medium-density"        , 0x20000000, 0x20000200, 0x20005000, 0x08000000, 0x08020000,  4, 1024 , 0x1FFFF000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x414, "STM32F High-density"          , 0x20000000, 0x20000200, 0x20010000, 0x08000000, 0x08080000,  2, 2048 , 0x1FFFF000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x418, "STM32F Connectivity line"     , 0x20000000, 0x20001000, 0x20010000, 0x08000000, 0x08040000,  2, 2048 , 0x1FFFB000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x420, "STM32F Low/Medium-density VL" , 0x20000000, 0x20000200, 0x20002000, 0x08000000, 0x08020000,  4, 1024 , 0x1FFFF000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x428, "STM32F High-density VL"       , 0x20000000, 0x20000200, 0x20008000, 0x08000000, 0x08080000,  2, 2048 , 0x1FFFF000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x430, "STM32F XL-density"            , 0x20000000, 0x20000800, 0x20018000, 0x08000000, 0x08100000,  2, 2048 , 0x1FFFE000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x416, "STM32L Medium-density"        , 0x20000000, 0x20000800, 0x20004000, 0x08000000, 0x08020000, 16,  256 , 0x1FF00000, 0x1FF01000, 0x1FF80000, 0x1FF8000F, 0x08080000, 0x08081000},
	{0x436, "STM32L High-density"          , 0x20000000, 0x20001000, 0x2000C000, 0x08000000, 0x08060000, 16,  256 , 0x1FF00000, 0x1FF02000, 0x1FF80000, 0x1FF8001F, 0x08080000, 0x08083000},
	{0x440, "STM32F051x"                   , 0x20000000, 0x20000800, 0x20002000, 0x08000000, 0x08010000,  4, 1024 , 0x1FFFEC00, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80B, 0x00000000, 0x00000000},
	/* Note that F2 and F4 devices have sectors of different page sizes
           and only the first sectors (of one page size) are included here */
	{0x411, "STM32F2xx"                    , 0x20000000, 0x20002000, 0x20020000, 0x08000000, 0x08100000,  4, 16384, 0x1FFF0000, 0x1FFF7800, 0x1FFFC000, 0x1FFFC00F, 0x00000000, 0x00000000},
	{0x413, "STM32F4xx"                    , 0x20000000, 0x20002000, 0x20020000, 0x08000000, 0x08100000,  4, 16384, 0x1FFF0000, 0x1FFF7800, 0x1FFFC000, 0x1FFFC00F, 0x00000000, 0x00000000},
	/* These are not (yet) in AN2606 - reserved by bootloader memory not known: */
	{0x427, "STM32L Medium-density Plus"   , 0x20000000, 0x20000800, 0x2000C000, 0x08000000, 0x08040000, 16,  256 , 0x1FF00000, 0x1FF02000, 0x1FF80000, 0x1FF8001F, 0x08080000, 0x08082000},
	{0x422, "STM32F30x & F31x"             , 0x20000000, 0x20002000, 0x20003000, 0x08000000, 0x08040000,  2, 2048 , 0x1FFFE000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x432, "STM32F37x & F38x"             , 0x20000000, 0x20002000, 0x20003000, 0x08000000, 0x08040000,  2, 2048 , 0x1FFFE000, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80F, 0x00000000, 0x00000000},
	{0x444, "STM32F050x"                   , 0x20000000, 0x20000800, 0x20001000, 0x08000000, 0x08008000,  4, 1024 , 0x1FFFEC00, 0x1FFFF800, 0x1FFFF800, 0x1FFFF80B, 0x00000000, 0x00000000},
	{0x0}
};

/* internal functions */
uint8_t stm32_gen_cs(const uint32_t v);
void    stm32_send_byte(const stm32_t *stm, uint8_t byte);
uint8_t stm32_read_byte(const stm32_t *stm);
char    stm32_send_command(const stm32_t *stm, const uint8_t cmd);


uint8_t stm32_gen_cs(const uint32_t v) {
	return  ((v & 0xFF000000) >> 24) ^
		((v & 0x00FF0000) >> 16) ^
		((v & 0x0000FF00) >>  8) ^
		((v & 0x000000FF) >>  0);
}

void stm32_send_byte(const stm32_t *stm, uint8_t byte) {
	serial_err_t err;
	err = serial_write(stm->serial, &byte, 1);
	if (err != SERIAL_ERR_OK) {
		fprintf(stderr, "Failed to send byte: ");
		perror("send_byte");
		exit(1);
	}
}

uint8_t stm32_read_byte(const stm32_t *stm) {
	uint8_t byte;
	serial_err_t err;
	err = serial_read(stm->serial, &byte, 1, NULL);
	if (err == SERIAL_ERR_NODATA) {
		fprintf(stderr, "Failed to read byte: read timeout\n");
		exit(1);
	} else if (err != SERIAL_ERR_OK) {
		fprintf(stderr, "Failed to read byte: ");
		perror("read_byte");
		exit(1);
	}
	return byte;
}

char stm32_send_command(const stm32_t *stm, const uint8_t cmd) {
	int ret;

	stm32_send_byte(stm, cmd);
	stm32_send_byte(stm, cmd ^ 0xFF);
	ret = stm32_read_byte(stm);
	if (ret == STM32_ACK) {
		return 1;
	} else if (ret == STM32_NACK) {
		fprintf(stderr, "Got NACK from device on command 0x%02x\n", cmd);
	} else {
		fprintf(stderr, "Unexpected reply from device on command 0x%02x\n", cmd);
	}
	return 0;
}

stm32_t* stm32_init(const serial_t *serial, const char init) {
	uint8_t len;
	stm32_t *stm;

	stm      = calloc(sizeof(stm32_t), 1);
	stm->cmd = calloc(sizeof(stm32_cmd_t), 1);
	stm->serial = serial;

	if (init) {
		uint8_t index;
		uint8_t ans = 0;
		serial_err_t err = 0;
		for(index = 5; index > 0; index--) {
			stm32_send_byte(stm, STM32_CMD_INIT);
			err = serial_read(stm->serial, &ans, 1, NULL);
			if (err == SERIAL_ERR_OK)
				break;
			if (err != SERIAL_ERR_NODATA) {
				fprintf(stderr, "Failed to read byte: ");
				perror("read_byte");
				exit(1);
			}
		}
		if(index == 0) {
			fprintf(stderr, "Failed to read byte: read timeout\n");
			exit(1);
		}
		if (ans == STM32_NACK) {
			fprintf(stderr, "Got NACK from INIT! Trying to resume connection...\n");
		} else if (ans != STM32_ACK) {
			stm32_close(stm);
			fprintf(stderr, "Failed to get init ACK (device return 0x%02X)\n", ans);
			return NULL;
		}
	}

	/* get the bootloader information */
	if (!stm32_send_command(stm, STM32_CMD_GET)) return 0;
	len              = stm32_read_byte(stm) + 1;
	stm->bl_version  = stm32_read_byte(stm); --len;
	stm->cmd->get    = stm32_read_byte(stm); --len;
	stm->cmd->gvr    = stm32_read_byte(stm); --len;
	stm->cmd->gid    = stm32_read_byte(stm); --len;
	stm->cmd->rm     = stm32_read_byte(stm); --len;
	stm->cmd->go     = stm32_read_byte(stm); --len;
	stm->cmd->wm     = stm32_read_byte(stm); --len;
	stm->cmd->er     = stm32_read_byte(stm); --len;
	stm->cmd->wp     = stm32_read_byte(stm); --len;
	stm->cmd->uw     = stm32_read_byte(stm); --len;
	stm->cmd->rp     = stm32_read_byte(stm); --len;
	stm->cmd->ur     = stm32_read_byte(stm); --len;
	if (len > 0) {
		fprintf(stderr, "Seems this bootloader returns more then we understand in the GET command, we will skip the unknown bytes\n");
		while(len-- > 0) stm32_read_byte(stm);
	}
	if (stm32_read_byte(stm) != STM32_ACK) {
		stm32_close(stm);
		return NULL;
	}

	/* get the version and read protection status  */
	if (!stm32_send_command(stm, stm->cmd->gvr)) {
		stm32_close(stm);
		return NULL;
	}

	stm->version = stm32_read_byte(stm);
	stm->option1 = stm32_read_byte(stm);
	stm->option2 = stm32_read_byte(stm);
	if (stm32_read_byte(stm) != STM32_ACK) {
		stm32_close(stm);
		return NULL;
	}

	/* get the device ID */
	if (!stm32_send_command(stm, stm->cmd->gid)) {
		stm32_close(stm);
		return NULL;
	}
	len = stm32_read_byte(stm) + 1;
	if (len < 2) {
		stm32_close(stm);
		fprintf(stderr, "Only %d bytes sent in the PID, unknown/unsupported device\n", len);
		return NULL;
	}
	stm->pid = (stm32_read_byte(stm) << 8) | stm32_read_byte(stm);
	len -= 2;
	if (len > 0) {
		fprintf(stderr, "This bootloader returns %d extra bytes in PID:", len);
		while (len-- > 0)
			fprintf(stderr, " %02x", stm32_read_byte(stm));
		fprintf(stderr, "\n");
	}
	if (stm32_read_byte(stm) != STM32_ACK) {
		stm32_close(stm);
		return NULL;
	}

	stm->dev = devices;
	while(stm->dev->id != 0x00 && stm->dev->id != stm->pid)
		++stm->dev;

	if (!stm->dev->id) {
		fprintf(stderr, "Unknown/unsupported device (Device ID: 0x%03x)\n", stm->pid);
		stm32_close(stm);
		return NULL;
	}

	return stm;
}

void stm32_close(stm32_t *stm) {
	if (stm) free(stm->cmd);
	free(stm);
}

char stm32_read_memory(const stm32_t *stm, uint32_t address, uint8_t data[], unsigned int len) {
	uint8_t cs;
	unsigned int i;
	assert(len > 0 && len < 257);

	/* must be 32bit aligned */
	assert(address % 4 == 0);

	address = be_u32      (address);
	cs      = stm32_gen_cs(address);

	if (!stm32_send_command(stm, stm->cmd->rm)) return 0;
	if (serial_write(stm->serial, &address, 4) != SERIAL_ERR_OK)
		return 0;

	stm32_send_byte(stm, cs);
	if (stm32_read_byte(stm) != STM32_ACK) return 0;

	i = len - 1;
	stm32_send_byte(stm, i);
	stm32_send_byte(stm, i ^ 0xFF);
	if (stm32_read_byte(stm) != STM32_ACK) return 0;

	if (serial_read(stm->serial, data, len, NULL) != SERIAL_ERR_OK)
		return 0;

	return 1;
}

char stm32_write_memory(const stm32_t *stm, uint32_t address, const uint8_t data[], unsigned int len) {
	uint8_t cs;
	unsigned int i;
	int c, extra;
	assert(len > 0 && len < 257);

	/* must be 32bit aligned */
	assert(address % 4 == 0);

	address = be_u32      (address);
	cs      = stm32_gen_cs(address);

	/* send the address and checksum */
	if (!stm32_send_command(stm, stm->cmd->wm)) return 0;
	if (serial_write(stm->serial, &address, 4) != SERIAL_ERR_OK)
		return 0;

	stm32_send_byte(stm, cs);
	if (stm32_read_byte(stm) != STM32_ACK) return 0;

	/* setup the cs and send the length */
	extra = len % 4;
	if(extra) extra = 4 - extra;
	cs = len - 1 + extra;
	stm32_send_byte(stm, cs);

	/* write the data and build the checksum */
	for(i = 0; i < len; ++i)
		cs ^= data[i];

	if (serial_write(stm->serial, data, len) != SERIAL_ERR_OK)
		return 0;

	/* write the alignment padding */
	for(c = 0; c < extra; ++c) {
		stm32_send_byte(stm, 0xFF);
		cs ^= 0xFF;
	}

	/* send the checksum */
	stm32_send_byte(stm, cs);
	return stm32_read_byte(stm) == STM32_ACK;
}

char stm32_wunprot_memory(const stm32_t *stm) {
	int ret;
	if (!stm32_send_command(stm, stm->cmd->uw)) return 0;
//Write unprotect should return two ACK bytes - one for command reception and one for command execution
	ret = stm32_read_byte(stm);
	if (ret == STM32_ACK) {
		return 1;
	} else if (ret == STM32_NACK) {
		fprintf(stderr, "Got NACK from device on flash write unprotecting\n");
	} else {
		fprintf(stderr, "Unexpected reply from device on flash write unprotecting\n");
	}
	return 0;
}

char stm32_runprot_memory  (const stm32_t *stm) {
	int ret;
	if (!stm32_send_command(stm, stm->cmd->ur)) return 0;
//Read unprotect should return two ACK bytes - one for command reception and one for command execution
	ret = stm32_read_byte(stm);
	if (ret == STM32_ACK) {
		return 1;
	} else if (ret == STM32_NACK) {
		fprintf(stderr, "Got NACK from device on flash read unprotecting\n");
	} else {
		fprintf(stderr, "Unexpected reply from device on flash read unprotecting\n");
	}
	return 0;
}

char stm32_rprot_memory(const stm32_t *stm) {
	int ret;
	if (!stm32_send_command(stm, stm->cmd->rp)) return 0;
//Read protect should return two ACK bytes - one for command reception and one for command execution
	ret = stm32_read_byte(stm);
	if (ret == STM32_ACK) {
		return 1;
	} else if (ret == STM32_NACK) {
		fprintf(stderr, "Got NACK from device on flash read protecting\n");
	} else {
		fprintf(stderr, "Unexpected reply from device on flash read protecting\n");
	}
	return 0;
}

char stm32_erase_memory(const stm32_t *stm, uint16_t spage, uint16_t pages) {
	char res;

	if (!pages)
		return 1;

	if (!stm32_send_command(stm, stm->cmd->er)) {
		fprintf(stderr, "Can't initiate chip erase!\n");
		return 0;
	}

	/* The erase command reported by the bootloader is either 0x43 or 0x44 */
	/* 0x44 is Extended Erase, a 2 byte based protocol and needs to be handled differently. */
	if (stm->cmd->er == STM32_CMD_EE) {
 		/* Not all chips using Extended Erase support mass erase */
 		/* Currently known as not supporting mass erase is the Ultra Low Power STM32L15xx range */
 		/* So if someone has not overridden the default, but uses one of these chips, take it out of */
 		/* mass erase mode, so it will be done page by page. This maximum might not be correct either! */
		if (stm->pid == 0x416 && pages == 0xFFFF)
		{
			spage = 0;
			pages = (stm->dev->fl_end - stm->dev->fl_start) / stm->dev->fl_ps; /* works for the STM32L152RB with 128Kb flash */
		}


		if (pages == 0xFFFF) {
			stm32_send_byte(stm, 0xFF);
			stm32_send_byte(stm, 0xFF); // 0xFFFF the magic number for mass erase
			stm32_send_byte(stm, 0x00); // 0x00 the XOR of those two bytes as a checksum
			if ((res = stm32_read_byte(stm)) != STM32_ACK) {
				fprintf(stderr, "Mass erase failed (devicw return 0x%02X). Try specifying the number of pages to be erased.\n", res);
				return 0;
			} else

			return 1;
		}

		uint16_t pg_num;
		uint8_t pg_byte;
 		uint8_t cs = ((pages-1) >> 8) ^ ((pages-1) & 0xFF);

 		stm32_send_byte(stm, (pages-1) >> 8); // Number of pages to be erased, two bytes, MSB first
 		stm32_send_byte(stm, (pages-1) & 0xFF);

 		for (pg_num = spage; pg_num < (pages + spage); pg_num++) {
 			pg_byte = pg_num >> 8;
 			cs ^= pg_byte;
 			stm32_send_byte(stm, pg_byte);
 			pg_byte = pg_num & 0xFF;
 			cs ^= pg_byte;
 			stm32_send_byte(stm, pg_byte);
 		}
 		stm32_send_byte(stm, cs);

 		if (stm32_read_byte(stm) != STM32_ACK) {
 			fprintf(stderr, "Page-by-page erase failed. Check the maximum pages your device supports.\n");
			return 0;
 		}

 		return 1;
	}

	/* And now the regular erase (0x43) for all other chips */
	if (pages == 0xFFFF) {
		return stm32_send_command(stm, 0xFF);
	} else {
		uint8_t cs = 0;
		uint8_t pg_num;
		stm32_send_byte(stm, pages-1);
		cs ^= (pages-1);
		for (pg_num = spage; pg_num < (pages + spage); pg_num++) {
			stm32_send_byte(stm, pg_num);
			cs ^= pg_num;
		}
		stm32_send_byte(stm, cs);
		return stm32_read_byte(stm) == STM32_ACK;
	}
}

char stm32_run_raw_code(const stm32_t *stm, uint32_t target_address, const uint8_t *code, uint32_t code_size)
{
	uint32_t stack_le = le_u32(0x20002000);
	uint32_t code_address_le = le_u32(target_address + 8);
	uint32_t length = code_size + 8;

	/* Must be 32-bit aligned */
	assert(target_address % 4 == 0);

	uint8_t *mem = malloc(length);
	if (!mem)
		return 0;

	memcpy(mem, &stack_le, sizeof(uint32_t));
	memcpy(mem + 4, &code_address_le, sizeof(uint32_t));
	memcpy(mem + 8, code, code_size);

	uint8_t *pos = mem;
	uint32_t address = target_address;
	while(length > 0) {

		uint32_t w = length > 256 ? 256 : length;
		if (!stm32_write_memory(stm, address, pos, w)) {
			free(mem);
			return 0;
		}

		address += w;
		pos += w;
		length -=w;
	}

	free(mem);
	return stm32_go(stm, target_address);
}

char stm32_go(const stm32_t *stm, uint32_t address) {
	uint8_t cs;

	address = be_u32      (address);
	cs      = stm32_gen_cs(address);

	if (!stm32_send_command(stm, stm->cmd->go)) return 0;
	serial_write(stm->serial, &address, 4);
	serial_write(stm->serial, &cs     , 1);

	return stm32_read_byte(stm) == STM32_ACK;
}

char stm32_reset_device(const stm32_t *stm) {
	uint32_t target_address = stm->dev->ram_bl_res;

	return stm32_run_raw_code(stm, target_address, stm_reset_code, stm_reset_code_length);
}

